hyppo/kgof/_utils.py

```diff
@@ -0,0 +1,310 @@
+    f = np.vectorize(each_elem_true)
+    return f(X)
+    
+
+def tr_te_indices(n, tr_proportion, seed=9282 ):
+    """Get two logical vectors for indexing train/test points.
+    Return (tr_ind, te_ind)
+    """
+    rand_state = np.random.get_state()
+    np.random.seed(seed)
+
+    Itr = np.zeros(n, dtype=bool)
+    tr_ind = np.random.choice(n, int(tr_proportion*n), replace=False)
+    Itr[tr_ind] = True
+    Ite = np.logical_not(Itr)
+
+    np.random.set_state(rand_state)
+    return (Itr, Ite)
+
+def subsample_ind(n, k, seed=32):
+    """
+    Return a list of indices to choose k out of n without replacement
+    """
+    with NumpySeedContext(seed=seed):
+        ind = np.random.choice(n, k, replace=False)
+    return ind
+
+def subsample_rows(X, k, seed=29):
+    """
+    Subsample k rows from the matrix X.
+    """
+    n = X.shape[0]
+    if k > n:
+        raise ValueError('k exceeds the number of rows.')
+    ind = subsample_ind(n, k, seed=seed)
+    return X[ind, :]
+    
+
+def fit_gaussian_draw(X, J, seed=28, reg=1e-7, eig_pow=1.0):
+    """
+    Fit a multivariate normal to the data X (n x d) and draw J points 
+    from the fit. 
+    - reg: regularizer to use with the covariance matrix
+    - eig_pow: raise eigenvalues of the covariance matrix to this power to construct 
+        a new covariance matrix before drawing samples. Useful to shrink the spread 
+        of the variance.
+    """
+    with NumpySeedContext(seed=seed):
+        d = X.shape[1]
+        mean_x = np.mean(X, 0)
+        cov_x = np.cov(X.T)
+        if d==1:
+            cov_x = np.array([[cov_x]])
+        [evals, evecs] = np.linalg.eig(cov_x)
+        evals = np.maximum(0, np.real(evals))
+        assert np.all(np.isfinite(evals))
+        evecs = np.real(evecs)
+        shrunk_cov = evecs.dot(np.diag(evals**eig_pow)).dot(evecs.T) + reg*np.eye(d)
+        V = np.random.multivariate_normal(mean_x, shrunk_cov, J)
+    return V
+
+def bound_by_data(Z, Data):
+    """
+    Determine lower and upper bound for each dimension from the Data, and project 
+    Z so that all points in Z live in the bounds.
+    Z: m x d 
+    Data: n x d
+    Return a projected Z of size m x d.
+    """
+    n, d = Z.shape
+    Low = np.min(Data, 0)
+    Up = np.max(Data, 0)
+    LowMat = np.repeat(Low[np.newaxis, :], n, axis=0)
+    UpMat = np.repeat(Up[np.newaxis, :], n, axis=0)
+
+    Z = np.maximum(LowMat, Z)
+    Z = np.minimum(UpMat, Z)
+    return Z
+
+
+def one_of_K_code(arr):
+    """
+    Make a one-of-K coding out of the numpy array.
+    For example, if arr = ([0, 1, 0, 2]), then return a 2d array of the form 
+     [[1, 0, 0], 
+      [0, 1, 0],
+      [1, 0, 0],
+      [0, 0, 1]]
+    """
+    U = np.unique(arr)
+    n = len(arr)
+    nu = len(U)
+    X = np.zeros((n, nu))
+    for i, u in enumerate(U):
+        Ii = np.where( np.abs(arr - u) < 1e-8 )
+        #ni = len(Ii)
+        X[Ii[0], i] = 1
+    return X
+
+def fullprint(*args, **kwargs):
```

Comment:
why is this necessary

Reply:
It isn't, has been removed.

Reply:
Was just to ensure a more well-formed error message when a non-finite data object is created but I'm now using a workaround instead.